library/core/src/ptr/non_null.rs

```diff
@@ -1829,4 +1856,103 @@ mod verify {
         let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
         let _ = NonNull::new(maybe_null_ptr);
     }
+
+    // pub const fn dangling() -> Self
+    #[kani::proof_for_contract(NonNull::dangling)]
+    pub fn non_null_check_dangling() {
+        // unsigned integer types
+        let ptr_u8 = NonNull::<u8>::dangling();
+        let ptr_u16 = NonNull::<u16>::dangling();
+        let ptr_u32 = NonNull::<u32>::dangling();
+        let ptr_u64 = NonNull::<u64>::dangling();
+        let ptr_u128 = NonNull::<u128>::dangling();
+        let ptr_usize = NonNull::<usize>::dangling();
+        // signed integer types
+        let ptr_i8 = NonNull::<i8>::dangling();
+        let ptr_i16 = NonNull::<i16>::dangling();
+        let ptr_i32 = NonNull::<i32>::dangling();
+        let ptr_i64 = NonNull::<i64>::dangling();
+        let ptr_i128 = NonNull::<i128>::dangling();
+        let ptr_isize = NonNull::<isize>::dangling();
+        // unit type
+        let ptr_unit = NonNull::<()>::dangling();
+        // zero length slice from dangling unit pointer
+        let zero_len_slice =  NonNull::slice_from_raw_parts(ptr_unit, 0);
+    }
+
+    // pub const fn from_raw_parts(data_pointer: NonNull<()>, metadata: <T as super::Pointee>::Metadata,) -> NonNull<T>
+    #[kani::proof_for_contract(NonNull::from_raw_parts)]
+    pub fn non_null_check_from_raw_parts() {
+
+        const ARR_LEN: usize = 10000;
+        // Create a non-deterministic array and its slice
+        let arr: [i8; ARR_LEN] = kani::any();
+        let arr_slice = &arr[..];  
```

Comment:
I changed this line to `let arr_slice = kani::slice::any_slice_of_array(&arr);` but then the following checks failure:
```
Check 24: ptr::non_null::verify::non_null_check_from_raw_parts.assertion.1
	 - Status: FAILURE
	 - Description: "slice content must preserve"
	 - Location: library/core/src/ptr/non_null.rs:1897:13 in function ptr::non_null::verify::non_null_check_from_raw_parts
	 
Check 40: ptr::non_null::NonNull::<[i8]>::as_ref::<'_>.safety_check.2
	 - Status: FAILURE
	 - Description: "dereference failure: pointer invalid"
	 - Location: library/core/src/ptr/non_null.rs:394:18 in function ptr::non_null::NonNull::<[i8]>::as_ref::<'_>
```
The failures are due to this line:
```rust
kani::assert( arr_slice == nonnull_slice.as_ref(), "slice content must preserve" );
```
Should I just remove the assertion?

Reply:
Why does the result pointer `nonnull_slice` become invalid? Is there any missing preconditions?

Reply:
@zhassan-aws helped identified the issue, I was still passing `ARR_LEN` which no longer accurate since the slice length is arbitrary now. Once I used `arr_slice.len()` the failures are resolved.